src/commands/AddCommand.js

```diff
@@ -0,0 +1,222 @@
+      .filter(a => this.packageSatisfied(a.name, a.versionRange))
+      .map(a => {
+        const pkg = this.getPackage(a.name);
+        return {
+          name: a.name,
+          version: pkg.version,
+          versionRange: a.versionRange
+        }
+      });
+
+    this.packagesToChange = this.filteredPackages
+      // Skip packages that only would install themselves
+      .filter(filteredPackage => {
+        const notSamePackage = pkgToInstall => pkgToInstall.name !== filteredPackage.name;
+        const addable = this.packagesToInstall.some(notSamePackage);
+        if (!addable) {
+          this.logger.warn(`Will not add ${filteredPackage.name} to itself.`);
+        }
+        return addable;
+      })
+      // Skip packages without actual changes to manifest
+      .filter(filteredPackage => {
+        const deps = filteredPackage[this.dependencyType] || {};
+
+        // Check if one of the packages to install necessiates a change to filteredPackage's manifest
+        return this.packagesToInstall
+          .filter(pkgToInstall => pkgToInstall.name !== filteredPackage.name)
+          .some(pkgToInstall => {
+            if (!(pkgToInstall.name in deps)) {
+              return true;
+            }
+            const current = deps[pkgToInstall.name];
+            const range = getRangeToReference(current, pkgToInstall.version, pkgToInstall.versionRange);
+            return range !== current;
+          });
+      });
+
+
+    if (this.packagesToChange.length === 0) {
+      const packagesToInstallList = this.packagesToInstall.map(pkg => pkg.name).join(', ');
+      this.logger.warn(`No packages found in scope where ${packagesToInstallList} can be added.`);
+    }
+
+    callback(null, this.packagesToChange.length > 0);
+  }
+
+  execute(callback) {
+    this.logger.info(`Add ${this.dependencyType} in ${this.packagesToChange.length} packages`);
+
+    this.packagesToChange
+      .forEach(pkgToChange => {
+        const manifestPath = path.join(pkgToChange.location, "package.json");
+
+        const notSamePackage = pkgToInstall => pkgToChange.name !== pkgToInstall.name;
+
+        const applicable = this.packagesToInstall
+          .filter(notSamePackage)
+          .reduce((results, pkgToInstall) => {
+            const deps = pkgToChange[this.dependencyType] || {};
+            const current = deps[pkgToInstall.name];
+            const range = getRangeToReference(
+              current,
+              pkgToInstall.version,
+              pkgToInstall.versionRange
+            );
+            const id = `${pkgToInstall.name}@${range}`;
+            const message = `Add ${id} as ${this.dependencyType} in ${pkgToChange.name}`;
+            this.logger.verbose(message);
+            results[pkgToInstall.name] = range;
+            return results;
+          }, {});
+
+        const source = FileSystemUtilities.readFileSync(manifestPath);
+
+        const file = new JSONFile(manifestPath, source);
```

Comment:
Is it really necessary to add `JSONFile`? You should be using `write-pkg`, an existing dependency, to write the package.json back to disk, as it already handles alphabetizing the updated (dev-)dependencies.

If you use `read-pkg` instead of reading the string, you can just merge `applicable` into the targeted `dependencyType` and write it back without testing, since the serialization is always alphabetized anyway.

```js
const manifestJson = readPkg.sync(manifestPath, { normalize: false });
const manifestDeps = manifestJson[dependencyType] || {};

// overwrite updated dependencies
manifestJson[dependencyType] = {
  ...manifestDeps,
  ...applicable,
};

writePkg.sync(manifestPath, manifestJson);
```

Reply:
I used `json-file-plus` to preserve `package.json` formatting. As this differs in behaviour from `npm` and other `lerna` commands and does not handle key alphabetization, I'll refactor as outlined.